bench_test.go

```diff
@@ -4,80 +4,121 @@
+		qt.Assert(b, err, qt.IsNil)
+		env := append(os.Environ(), "GOGARBLE=*", "GOCACHE="+gocache)
+		args := []string{"build", "-v", "-o=" + outputBin, sourceDir}
+
+		cmd := exec.Command(garbleBin, args...)
+		cmd.Env = env
+		cmd.Dir = sourceDir
+
+		out, err := cmd.CombinedOutput()
+		qt.Assert(b, err, qt.IsNil, qt.Commentf("output: %s", out))
+		// Ensure that we built all packages, as expected.
+		qt.Assert(b, rxBuiltRuntime.Match(out), qt.IsTrue)
+		qt.Assert(b, rxBuiltMain.Match(out), qt.IsTrue)
 
-	for _, name := range [...]string{"Cache", "NoCache"} {
-		b.Run(name, func(b *testing.B) {
-			buildArgs := []string{"build", "-o=" + b.TempDir()}
-			switch name {
-			case "Cache":
-				buildArgs = append(buildArgs, "./testdata/bench-cache")
-
-				// Ensure the build cache is warm,
-				// for the sake of consistent results.
-				cmd := exec.Command(garbleBin, buildArgs...)
-				if out, err := cmd.CombinedOutput(); err != nil {
-					b.Fatalf("%v: %s", err, out)
-				}
-			case "NoCache":
-				buildArgs = append(buildArgs, "./testdata/bench-nocache")
-			default:
-				b.Fatalf("unknown name: %q", name)
-			}
-
-			// We collect extra metrics.
-			var userTime, systemTime int64
-
-			b.ResetTimer()
-			b.RunParallel(func(pb *testing.PB) {
-				for pb.Next() {
-					cmd := exec.Command(garbleBin, buildArgs...)
-					if name == "NoCache" {
-						gocache, err := os.MkdirTemp(b.TempDir(), "gocache-*")
-						if err != nil {
-							b.Fatal(err)
-						}
-						cmd.Env = append(os.Environ(), "GOCACHE="+gocache)
-					}
-					if out, err := cmd.CombinedOutput(); err != nil {
-						b.Fatalf("%v: %s", err, out)
-					}
-
-					userTime += int64(cmd.ProcessState.UserTime())
-					systemTime += int64(cmd.ProcessState.SystemTime())
-				}
-			})
-			b.ReportMetric(float64(userTime)/float64(b.N), "user-ns/op")
-			b.ReportMetric(float64(systemTime)/float64(b.N), "sys-ns/op")
-			info, err := os.Stat(garbleBin)
-			if err != nil {
-				b.Fatal(err)
-			}
-			b.ReportMetric(float64(info.Size()), "bin-B")
-		})
+		// The cached rebuild will reuse all dependencies,
+		// but rebuild the main package itself.
+		cachedStart := time.Now()
+		writeSourceFile("rebuild.go", []byte(fmt.Sprintf("package main\nvar v%d int", i)))
+
+		cmd = exec.Command(garbleBin, args...)
+		cmd.Env = env
+		cmd.Dir = sourceDir
+
+		out, err = cmd.CombinedOutput()
+		qt.Assert(b, err, qt.IsNil, qt.Commentf("output: %s", out))
+		// Ensure that we only rebuilt the main package, as expected.
+		qt.Assert(b, rxBuiltRuntime.Match(out), qt.IsFalse)
+		qt.Assert(b, rxBuiltMain.Match(out), qt.IsTrue)
+
+		userTime += int64(cmd.ProcessState.UserTime())
+		systemTime += int64(cmd.ProcessState.SystemTime())
+		cachedTime += time.Since(cachedStart).Nanoseconds()
+	}
+	b.ReportMetric(float64(cachedTime)/float64(b.N), "cached-ns/op")
+	b.ReportMetric(float64(userTime)/float64(b.N), "user-ns/op")
+	b.ReportMetric(float64(systemTime)/float64(b.N), "sys-ns/op")
```

Comment:
Same order

Reply:
I've changed the other two bits of code to put "cached time" first. I think keeping it first makes sense, because as a wall-time metric it's more akin to ns/op than the others. It doesn't really matter with the results, because I think they are sorted.